frame/uniswap-v2/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+			PoolAssetBalance::<T>::mutate(pool_id, asset_id_j, |balance| -> DispatchResult {
+				*balance = bal_j.checked_mul_int(1u64).ok_or(Error::<T>::Math)?.into();
+				Ok(())
+			})?;
+
+			ensure!(T::LpToken::balance(asset_id_i, &who) >= dx, Error::<T>::InsufficientFunds);
+
+			ensure!(
+				T::LpToken::balance(asset_id_j, &Self::account_id(&pool_id)) >= dy,
+				Error::<T>::InsufficientFunds
+			);
+
+			Self::transfer_liquidity_into_pool(&Self::account_id(&pool_id), pool_id, &who, i, dx)?;
+			Self::transfer_liquidity_from_pool(&Self::account_id(&pool_id), pool_id, j, &who, dy)?;
+
+			Self::deposit_event(Event::TokenExchanged {
+				who: who.clone(),
+				pool_id,
+				sent_token_index: i_token,
+				sent_amount: dx,
+				received_token_index: j_token,
+				received_amount: dy,
+				fee: dy_fee,
+			});
+			Ok(())
+		}
+
+		fn withdraw_admin_fees(
+			who: &Self::AccountId,
+			pool_id: PoolId,
+			admin_fee_account: &Self::AccountId,
+		) -> Result<(), DispatchError> {
+			let assets = PoolAssets::<T>::get(pool_id).ok_or(Error::<T>::InconsistentStorage)?;
+			let mut balances = Vec::new();
+			for asset_id in &assets {
+				balances.push(PoolAssetBalance::<T>::get(pool_id, asset_id));
+			}
+			let mut total_balances = Vec::new();
+			for asset_id in &assets {
+				total_balances.push(PoolAssetTotalBalance::<T>::get(pool_id, asset_id));
+			}
+			let n_coins = assets.len();
+
+			ensure!(n_coins == balances.len(), Error::<T>::InconsistentStorage);
+
+			let admin_fees = total_balances
+				.iter()
+				.zip(balances)
+				.map(|(tb, b)| {
+					let admin_fee = tb.checked_sub(&b).ok_or(Error::<T>::Math)?;
+
+					Ok(admin_fee)
+				})
+				.collect::<Result<Vec<Self::Balance>, DispatchError>>()?;
+
+			let assets = assets.clone();
+
+			for (index, asset_id) in assets.iter().enumerate() {
+				let new_total_balance = total_balances[index]
+					.checked_sub(&admin_fees[index])
+					.ok_or(Error::<T>::Math)?;
+				PoolAssetTotalBalance::<T>::mutate(
+					pool_id,
+					asset_id,
+					|total_balance| -> DispatchResult {
+						*total_balance = new_total_balance;
+						Ok(())
+					},
+				)?;
+			}
+
+			for (asset, amount) in assets.into_iter().zip(admin_fees.iter().copied()) {
+				T::LpToken::transfer(
+					asset,
+					&Self::account_id(&pool_id),
+					admin_fee_account,
+					amount,
+					true,
+				)?;
+			}
+
+			Self::deposit_event(Event::AdminFeesWithdrawn {
+				who: who.clone(),
+				pool_id,
+				admin_fee_account: admin_fee_account.clone(),
+				admin_fees,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Return pool information for given pool_id.
+		pub fn get_pool_info(pool_id: PoolId) -> Option<ConstantProductPoolInfo<T::AccountId>> {
+			let pool_data = Pools::<T>::get(pool_id)?;
+			let pool;
+			match pool_data {
+				PoolData::ConstantProductPoolData(pool_info) => {
+					pool = pool_info;
+				},
+				_ => {
+					panic!("Only constant product pool expected.");
```

Comment:
not sure why it is needed here

Reply:
Need to handle all variants of Enum right?